x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java

```diff
@@ -285,6 +285,111 @@ public void onFailure(Exception e) {
         }
     }
 
+    public void deleteRoles(
+        final List<String> roleNames,
+        WriteRequest.RefreshPolicy refreshPolicy,
+        final ActionListener<BulkRolesResponse> listener
+    ) {
+        if (enabled == false) {
+            listener.onFailure(new IllegalStateException("Native role management is disabled"));
+            return;
+        }
+
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        Map<String, Exception> validationErrorByRoleName = new HashMap<>();
+
+        for (String roleName : roleNames) {
+            if (reservedRoleNameChecker.isReserved(roleName)) {
+                validationErrorByRoleName.put(
+                    roleName,
+                    new IllegalArgumentException("role [" + roleName + "] is reserved and cannot be deleted")
+                );
+            } else {
+                bulkRequest.add(createRoleDeleteRequest(roleName));
+            }
+        }
+
+        if (bulkRequest.numberOfActions() == 0) {
+            BulkRolesResponse.Builder bulkDeleteRolesResponseBuilder = new BulkRolesResponse.Builder();
+            roleNames.stream()
+                .map(roleName -> BulkRolesResponse.Item.failure(roleName, validationErrorByRoleName.get(roleName)))
+                .forEach(bulkDeleteRolesResponseBuilder::addItem);
+
+            listener.onResponse(bulkDeleteRolesResponseBuilder.build());
+            return;
+        }
+
+        final SecurityIndexManager frozenSecurityIndex = securityIndex.defensiveCopy();
+        if (frozenSecurityIndex.indexExists() == false) {
+            BulkRolesResponse.Builder bulkDeleteRolesResponseBuilder = new BulkRolesResponse.Builder();
+            roleNames.stream()
+                .map(
+                    roleName -> BulkRolesResponse.Item.failure(roleName, new IllegalArgumentException("Role [" + roleName + "] not found"))
```

Comment:
can you please mention in that the .security index does not exist, and ideally make the exception not an `IllegalArgumentException` but rather an `ElasticsearchStattusException` with `RestStatus.SERVICE_UNAVAILABLE`?

Reply:
I was going back and forth on if this should be a per role error or a top level error. If we use `RestStatus.SERVICE_UNAVAILABLE` it feels like it should be a top level error. So a 400 `RestStatus.SERVICE_UNAVAILABLE`, instead of a 200 with individual errors per role. What do you think?

Reply:
> I was going back and forth on if this should be a per role error or a top level error. 

The dogma in the other (hopefully most) security APIs that "get" and "delete" (but not write), when the .security index does not exist, is to return a 200, as if the index exists but it is empty. The .security index is created lazily, when we add anything to it so everything "feels nice and consistent to users" (they don't have to ensure that any "get" must in all cases follow a "put", otherwise they get an error).
I remember us discussing to change this, but I don't recall more context. I think a problem is that "the index does not exist" here covers also the case where the "index has not been recovered yet".

>  So a 500 RestStatus.SERVICE_UNAVAILABLE, instead of a 200 with individual errors per role. What do you think?

Given the prevailing convention of not erroring on "index not existing", I think for bulk APIs we should especially follow that convention, since, in addition, the response for bulk should still be 200 even if some individual requests failed (and by extension if all of them failed).

So I think, for bulk delete, the overall error code should be 200, with individual 503 failures, when the .security index does not exist (or the state is not recovered).

Reply:
Thanks for the explanation, that makes sense. I'll update it to throw an individual service unavailable instead. 👍

Reply:
Looking some more it looks like the convention is to return an empty list (no error). [NativePrivilegeStore](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java#L458), [ApiKeyService](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1492C1-L1493C1)

So I've changed it to return an empty list of results instead. Let me know what you think.

Reply:
Ok lets go with that. I still feel that in those cases we didn't have a better option to signal that the .security index does not exist AND still return a 200. But in this case we do have some way.

But again, I need we need to have a broader discussion on what to do when the index does not exist. I'll set that up. For now, I'm OK to return an empty response. We can change to add errors for it later (more easily than remove errors we would be adding now).